README.md
éNuage version updater command
=======================

[![Packagist](https://img.shields.io/packagist/v/enuage/version-updater.svg)](https://packagist.org/packages/enuage/version-updater)
[![Packagist](https://img.shields.io/packagist/l/enuage/version-updater.svg)](https://packagist.org/packages/enuage/version-updater)

Requirements
============

- PHP >= 7.1.3
- Symfony is optional

Installation
============

Applications that don't use Symfony at all
----------------------------------

### Step 1: Download the library

Open a command console, enter your project directory and execute:

```console
$ composer require enuage/version-updater
```

### Step 2: Use the command for version updating from `vendor/bin`

```
$ php vendor/bin/version-updater [options] [--] [<version>]
```

Applications that use Symfony Flex
----------------------------------

Open a command console, enter your project directory and execute:

```console
$ composer require enuage/version-updater
```

Applications that don't use Symfony Flex
----------------------------------------

### Step 1: Download the Bundle

Open a command console, enter your project directory and execute the
following command to download the latest stable version of this bundle:

```console
$ composer require version-updater
```

This command requires you to have Composer installed globally, as explained
in the [installation chapter](https://getcomposer.org/doc/00-intro.md)
of the Composer documentation.

### Step 2: Enable the Bundle

Then, enable the bundle by adding it to the list of registered bundles
in the `app/AppKernel.php` file of your project:

```php
<?php
// app/AppKernel.php

// ...
class AppKernel extends Kernel
{
    public function registerBundles()
    {
        $bundles = array(
            // ...
            new Enuage\VersionUpdaterBundle\VersionUpdaterBundle(),
        );
        // ...
    }
    // ...
}
```

Usage
=====

## Option 1: Updating files via command line

### Step 1: Define files and regular expression for version updating

Use `\V` for define the version in regular expression.
It will be replaced with the SemVer regular expression

**New in version 1.2.0**: you can easily define property of the json
file which should be updated - just add file path and path to the
property delimited by slash under `json` configuration in following
format: `path-to/file: property/path`

**New in version 1.3.0**: you can easily define property of the yaml
file which should be updated - just add file path and path to the
property delimited by slash under `yaml` configuration in following
format: `path-to/file: property/path`

**New in version 1.5.4**: custom commit and tag (release) messages

Example:

```yaml
files:
  - '.env': '/^(API_VERSION=)\V/m'
  - 'README.md': '/^(Version:\s)\V/m'
json:
  - composer: version                 # File: <project>/composer.json
  - doc/api: info/version             # File <project>/doc/api.json
yaml:
  - doc/api: info/version             # File <project>/doc/api.yaml
git:
  enabled: true # Optional parameter. You can disable Git operations without removing its configuration
  source: true # The latest tag will be got from Git history
  push: true # Push tag to repository after its creation
  prefix: v # Tag prefix. Empty by default
  message:
    commit: 'Your custom commit message (\V)' # If you want, you can use \V to define the place for version string
    release: 'Your custom release message'
```

### Step 2: Use the command for version updating

```
$ php bin/console enuage:version:update [options] [--] [<version>]
```

Available options
-----------------

- `--major`: Increase only major version
- `--minor`: Increase only minor version
- `--patch`: Increase only patch version
- `--alpha`: Increase or define the alpha version
- `--beta`: Increase or define the beta version
- `--rc`: Increase or define the release candidate version
- `--down`: Decrease defined version. It's also applicable to prerelease versions
- `--release`: Remove all prerelease versions
- `--date <none|PHP date format>`: Add date metadata to the version. By default date format is equal to `c`
- `--meta <data>`: Add metadata to the version
- `--config-file <string>`: Path to configurations file or to directory with `.enuage` configuration file
- `--colors <boolean>`: Enable/disable colors in output. Default - `true`
- `--show-current <string|boolean>`: Show current version from file or from all available sources. Currently supported
sources:
    - composer
    - Git repository (if enabled)
- `--exclude-git`: Disable updating Git repository
- `--print-version-only`: Hide all messages and print only new version when script job will be finished. Messages can be
displayed again with increased verbosity level (`-v`)

## Option 2: Updating version via service

Example:

```php
use Enuage\VersionUpdaterBundle\DTO\VersionOptions;
use Symfony\Component\DependencyInjection\Container;
// use Enuage\VersionUpdaterBundle\Service\VersionService;

$version = '0.1.0-alpha.2';

$service = $container->get('enuage.version.service');
// $service = VersionService(); // If you don't use Symfony
$options = new VersionOptions();

$options->increasePreRelease();

$service->update($version, $options); // Result: "0.1.0-alpha.3"
```

Available methods
-----------------

- `addDateMeta(format = null)`: Enable date meta in provided format or ['c'][1] by default
- `addMeta(value = null)`: Add custom meta to the tag
- `increaseMajor()`: Enable major version increase
- `decreaseMajor()`: Disable major version increase
- `increaseMinor()`: Enable minor version increase
- `decreaseMinor()`: Disable minor version increase
- `increasePatch()`: Enable patch version increase
- `decreasePatch()`: Disable patch version increase
- `updateAlpha()`: Enable pre-release version `alpha` modifications
- `updateBeta()`: Enable pre-release version `beta` modifications
- `updateReleaseCandidate()`: Enable pre-release version `rc` modifications
- `increasePreRelease()`: Increase pre-release version  (e.g: `0.1.0-alpha.2` -> `0.1.0-alpha.3`)
- `decreasePreRelease()`: Decrease pre-release version (e.g: `0.1.0-alpha.2` -> `0.1.0-alpha.1`)
- `setVersion()`: Set custom version value
- `downgrade()`: Downgrade the version, including pre-release
- `release()`: Disable all pre-release postfixes

[1]: https://www.php.net/manual/en/function.date.php


CHANGELOG.md
# Version updater command changelog

### v1.5.3

- Fix bugs

### v1.5.2

- Fix `bin` property in composer configuration file

### v1.5.1

- Update composer requirements
- Add new option for hide all messages except new version

### v1.5.0

- [EVUB-7][3]: Added Git tag updating and parsing latest version from remote
repository

### v1.4.1

- Added feature for disabling colors in output

### v1.4.0

- [EVUB-4][4]: Added executable file for using command from `vendor/bin` directory
- Updated files finder
- Updated command
    - improved output
    - add debug verbosity
- Improved and cleaned code
- Add configurations parser
- Update handlers and files formatter
- Update helpers
    - StringType improvements
    - Added BooleanType

### v1.3.1

##### 2019-07-06 00:28

- Fixed possible bug with `.yml` extension
- Code improvements

### v1.3.0

##### 2019-06-30 09:51

- [EVUB-2][2]: added handler for YAML files version updating by pattern

### v1.2.0

##### 2019-06-30 05:48

- [EVUB-1][1]: added handler for JSON files version updating by pattern

### v1.1.0

##### 2019-06-30 22:22

- Enable GitLab CI
- Update documentation
- Add version parser test
- Update code of pre-release versions data manipulations
- Code clean and improvements

##### 2019-06-29

- Fix bug on decreasing minor and patch version at the same time

##### 2019-06-23 17:36

- Add `command options parser`
- Add `files finder`, moved all files operations there
- Add `files array normalizer`
- Add `version options` DTO
- Updated `mutator`, moved there all code from command
- Add tests for mutator (Coverage 99.72%: 34 tests, 102 assertions) 

##### 2019-06-23 02:06

- Code clean and improvements
- Added prefix recognition
- Large code refactoring. Moved the code from command to separate
files

### v1.0.2

##### 2018-12-20 02:23

- Fix bugs
- Add tests (Coverage 100%: 26 tests, 65 assertions)

### v1.0.1

##### 2018-12-05 10:24:43

- Fix code

### v1.0.0

##### 2018-12-03 01:46:

- Version 1.0.0 release

### v0.1.1-alpha

##### 2018-12-02 22:35:

- Fixed cache issue

### v0.1.0-alpha

##### 2018-12-02 21:14:

- Moved version updater command from API project to separated bundle

[1]: https://enuage.myjetbrains.com/youtrack/issue/EVUB-1
[2]: https://enuage.myjetbrains.com/youtrack/issue/EVUB-2
[3]: https://enuage.myjetbrains.com/youtrack/issue/EVUB-7
[4]: https://enuage.myjetbrains.com/youtrack/issue/EVUB-4


src/ValueObject/Version.php
<?php
/**
 * Version
 *
 * Created at 2019-06-22 12:26 AM
 *
 * @author Serghei Niculaev <spam312sn@gmail.com>
 * @license GNU GPLv3 <https://www.gnu.org/licenses/gpl-3.0.en.html>
 *
 * This file is a part of éNuage version updater command
 *
 * For the full copyright and license information, please view the LICENSE.md
 * file that was distributed with this source code.
 */

namespace Enuage\VersionUpdaterBundle\ValueObject;

use Enuage\VersionUpdaterBundle\Collection\ArrayCollection;
use Enuage\VersionUpdaterBundle\Collection\VersionComponentsCollection;

/**
 * Class Version
 *
 * @author Serghei Niculaev <spam312sn@gmail.com>
 */
class Version
{
    public const MAJOR = 'major';
    public const MINOR = 'minor';
    public const PATCH = 'patch';

    public const MAIN_VERSIONS = [
        self::MAJOR,
        self::MINOR,
        self::PATCH,
    ];

    public const ALPHA = 'alpha';
    public const BETA = 'beta';
    public const RELEASE_CANDIDATE = 'rc';

    public const PRE_RELEASE_VERSIONS = [
        self::ALPHA,
        self::BETA,
        self::RELEASE_CANDIDATE,
    ];

    public const META = 'meta';
    public const META_DATE = 'date';

    /**
     * @var string|null
     */
    private $prefix;

    /**
     * @var VersionComponentsCollection
     */
    private $mainComponents;

    /**
     * @var VersionComponentsCollection
     */
    private $preReleaseComponents;

    /**
     * @var ArrayCollection
     */
    private $metaComponents;

    /**
     * Version constructor.
     */
    public function __construct()
    {
        $this->mainComponents = new VersionComponentsCollection(self::MAIN_VERSIONS);
        $this->preReleaseComponents = new VersionComponentsCollection(
            self::PRE_RELEASE_VERSIONS,
            VersionComponentsCollection::DISABLE_ALL
        );

        $this->metaComponents = new ArrayCollection();
    }

    /**
     * @return null|string
     */
    public function getPrefix(): ?string
    {
        return $this->prefix;
    }

    /**
     * @param string $value
     *
     * @return Version
     */
    public function setPrefix(string $value): Version
    {
        $this->prefix = $value;

        return $this;
    }

    /**
     * @param string $type
     *
     * @return null|int
     */
    public function getMainVersion(string $type): ?int
    {
        return $this->getMainComponentValue($type);
    }

    /**
     * @param string $type
     *
     * @return int
     */
    private function getMainComponentValue(string $type): int
    {
        return $this->getMainComponents()->get($type)->getValue();
    }

    /**
     * @return VersionComponentsCollection
     */
    public function getMainComponents(): VersionComponentsCollection
    {
        return $this->mainComponents;
    }

    /**
     * @param string $type
     * @param int $value
     *
     * @return Version
     */
    public function setMainVersion(string $type, int $value): Version
    {
        $this->mainComponents->set($type, $value);

        return $this;
    }

    /**
     * @return int
     */
    public function getMajor(): int
    {
        return $this->getMainComponentValue(self::MAJOR);
    }

    /**
     * @param int $value
     *
     * @return Version
     */
    public function setMajor(int $value): Version
    {
        $this->mainComponents->set(self::MAJOR, $value);

        return $this;
    }

    /**
     * @return int
     */
    public function getMinor(): int
    {
        return $this->getMainComponentValue(self::MINOR);
    }

    /**
     * @param int $value
     *
     * @return Version
     */
    public function setMinor(int $value): Version
    {
        $this->mainComponents->set(self::MINOR, $value);

        return $this;
    }

    /**
     * @return int
     */
    public function getPatch(): int
    {
        return $this->getMainComponentValue(self::PATCH);
    }

    /**
     * @param int $value
     *
     * @return Version
     */
    public function setPatch(int $value): Version
    {
        $this->mainComponents->set(self::PATCH, $value);

        return $this;
    }

    /**
     * @return Version
     */
    public function clearPreRelease(): Version
    {
        $this->preReleaseComponents->disableAll();

        return $this;
    }

    /**
     * @param string $type
     *
     * @return Version
     */
    public function enablePreRelease(string $type): Version
    {
        $this->getPreReleaseComponent($type)->setEnabled(true);

        return $this;
    }

    /**
     * @param string $type
     *
     * @return VersionComponent
     */
    public function getPreReleaseComponent(string $type): VersionComponent
    {
        return $this->preReleaseComponents->getValue($type);
    }

    /**
     * @return null|int
     */
    public function getPreReleaseVersion(): ?int
    {
        return $this->getPreReleaseComponent($this->getPreRelease())->getValue();
    }

    /**
     * @return null|string
     */
    public function getPreRelease(): ?string
    {
        /** @var VersionComponent $component */
        foreach ($this->preReleaseComponents->getIterator() as $type => $component) {
            if ($component->isEnabled()) {
                return $type;
            }
        }

        return null;
    }

    /**
     * @param string $type
     * @param int $value
     */
    public function setPreReleaseVersion(string $type, int $value): void
    {
        $this->getPreReleaseComponent($type)->setValue($value);
    }

    /**
     * @return ArrayCollection
     */
    public function getMetaComponents(): ArrayCollection
    {
        return $this->metaComponents;
    }

    /**
     * @param ArrayCollection $metaComponents
     */
    public function setMetaComponents(ArrayCollection $metaComponents): void
    {
        $this->metaComponents = $metaComponents;
    }

    /**
     * @return string
     */
    public function implodeMainComponents(): string
    {
        return implode('.', $this->getMainComponents()->toArray());
    }
}
